### DISCUSSION TITLE
migration will not revert if it went wrong

### DISCUSSION THREAD

**Discussion by cdwcgt** [ppy/osu#16793] [Q&A] [answered]
![image](https://user-images.githubusercontent.com/34775378/152666307-8cba5b55-ce22-45ee-8ea9-cf1722227ec3.png)
It move some of my lazer file from `%AppData%\osu` to target folder
but when it went wrong, lazer will not revert it and will stay at `AppData`

![image](https://user-images.githubusercontent.com/34775378/152666337-5fe265e2-591c-4e47-a305-d2617a11f273.png)
fill access denied because osulazer is using it

I actually moved the folder very early on as well
after update to [2022.205.0](https://osu.ppy.sh/home/changelog/lazer/2022.205.0) and it try to update my database(maybe? I found it at https://github.com/ppy/osu/pull/16702). and it moved all my lazer file to appdata. So today I get error when I try to move it

**Comment by peppy** [comment, marked as answer]
Try restarting your PC if the file is access denied. Something might have it locked.

If it still doesn't work, please provide log files.

**Reply by cdwcgt** [reply]
the file is ruleset dll.
and I think the reason he is locked is that it is occupied by the lazer itself.

and seens like lazer delete my data after failed the migration so i may cannot found log

I will try to reproduce

**Reply by cdwcgt** [reply]
[logs.zip](https://github.com/ppy/osu/files/8016744/logs.zip)
here's log file

lazer move his work folder to target folder, but will go back to the previous directory after restart

**Reply by peppy** [reply]
I don't understand. Why is a folder migration running on startup?

**Reply by cdwcgt** [reply]
I am running after startup? but immediately

wait a second I will do it again

**Reply by peppy** [reply]
I think your best bet is to unload the ruleset and perform migration for now.

**Reply by cdwcgt** [reply]
ok I know why log lost startup section
osulazer delete all the file in the previous folder, also contain the log before migration error

and the log I send is after error

**Reply by cdwcgt** [reply]
> I think your best bet is to unload the ruleset and perform migration for now.

I actually solved the problem this way too
But I think can lazer write before migration
or go restart and perform migration

**Reply by peppy** [reply]
Rulesets likely shouldn't be in the user data folder. We may look to move them in the future.

**Reply by cdwcgt** [reply]
ok. thanks for paying attention to this issue :)

**Reply by cdwcgt** [reply]
but I think it just solve the problem that cause migration error
Is it possible to provide pause or revert for migration if error for other reasons?

no technical ability people may lose all data in lazer if migration goes error.

**Reply by peppy** [reply]
Interestingly, in testing this it does look like the copying of files to the new location succeeds correctly, but the operation fails on the delete stage. Can you confirm that all rulesets (and other files) are copied correctly?

**Reply by cdwcgt** [reply]
yes
it failed at delete section and rulesets are already copied

**Reply by peppy** [reply]
Tracking via https://github.com/ppy/osu/issues/16841.
